### PR TITLE
[DPE-5117] Sharding + TLS on MongoBD K8S

### PR DESCRIPTION
**Pull request by Gu1nness**
## Problem

 * Ensure Sharding + TLS is working

## Solution

 * Add test (copy-pasted + small fixes from MongoDB VM)
 * Fix small bugs in charm

### CODE REVIEW

File: tests/integration/tls_tests/test_tls.py [hunk 1]
```diff
@@ -1,85 +1,34 @@
...
 TLS_TEST_DATA = "tests/integration/tls_tests/data"
-EXTERNAL_CERT_PATH = "/etc/mongod/external-ca.crt"
-INTERNAL_CERT_PATH = "/etc/mongod/internal-ca.crt"
 DB_SERVICE = "mongod.service"
 
 logger = logging.getLogger(__name__)
 
 
-@pytest.mark.group(1)
-async def check_certs_correctly_distributed(
```
Comment (Gu1nness) [inline tests/integration/tls_tests/test_tls.py:40]: Moved to `tests/integration/tls_tests/helpers.py`

File: tests/integration/tls_tests/helpers.py [hunk 2]
```diff
@@ -113,6 +142,13 @@ async def scp_file_preserve_ctime(ops_test: OpsTest, unit_name: str, path: str)
     return f"{filename}"
 
 
+async def get_file_content(ops_test: OpsTest, unit_name: str, path: str) -> str:
```
Comment (Gu1nness) [inline tests/integration/tls_tests/helpers.py:145]: Need this function

File: tests/integration/tls_tests/helpers.py [hunk 3]
```diff
@@ -50,22 +68,33 @@ async def check_tls(ops_test: OpsTest, unit: ops.model.Unit, enabled: bool) -> b
...
-def parse_hostname(hostname: str) -> str:
+def parse_hostname(hostname: str, app_name: str) -> str:
     """Parses a hostname to the corresponding endpoint version."""
-    return hostname.replace("/", "-") + ".mongodb-k8s-endpoints"
+    unit_id = hostname.split("/")[1]
+    return f"{app_name}-{unit_id}.{app_name}-endpoints"
 
 
-async def run_tls_check(ops_test: OpsTest, unit: ops.model.Unit) -> int:
+async def run_tls_check(
```
Comment (Gu1nness) [inline tests/integration/tls_tests/helpers.py:77]: Update tls check to be sharding compatible

File: tests/integration/tls_tests/helpers.py [hunk 4]
```diff
@@ -50,22 +68,33 @@ async def check_tls(ops_test: OpsTest, unit: ops.model.Unit, enabled: bool) -> b
         return False
 
 
-def parse_hostname(hostname: str) -> str:
+def parse_hostname(hostname: str, app_name: str) -> str:
```
Comment (Gu1nness) [inline tests/integration/tls_tests/helpers.py:71]: Sharding compatible `parse_hostname`

File: tests/integration/tls_tests/helpers.py [hunk 5]
```diff
@@ -5,25 +5,43 @@
...
+INTERNAL_CERT_PATH = "/etc/mongod/internal-ca.crt"
+TLS_RELATION_NAME = "certificates"
 
 
 class ProcessError(Exception):
     """Raised when a process fails."""
 
 
-async def check_tls(ops_test: OpsTest, unit: ops.model.Unit, enabled: bool) -> bool:
+async def check_tls(
```
Comment (Gu1nness) [inline tests/integration/tls_tests/helpers.py:38]: Sharding compatible `check_tls`

File: tests/integration/sharding_tests/test_sharding_tls.py
Comment (Gu1nness) [inline tests/integration/sharding_tests/test_sharding_tls.py:1]: Copied from MongoDB VM with minor changes

File: tests/integration/sharding_tests/helpers.py [hunk 6]
```diff
@@ -1,8 +1,87 @@
...
+SHARD_TWO_APP_NAME = "shard-two"
+CONFIG_SERVER_APP_NAME = "config-server"
+CONFIG_SERVER_REL_NAME = "config-server"
+MONGODB_CHARM_NAME = "mongodb-k8s"
+SHARD_REL_NAME = "sharding"
+CLUSTER_COMPONENTS = [SHARD_ONE_APP_NAME, SHARD_TWO_APP_NAME, CONFIG_SERVER_APP_NAME]
+TIMEOUT = 15 * 60
+
+
+async def deploy_cluster_components(
```
Comment (Gu1nness) [inline tests/integration/sharding_tests/helpers.py:20]: Copied from MongoDB VM + minor changes (add resources)

File: src/charm.py [hunk 7]
```diff
@@ -1191,6 +1206,9 @@ def restart_charm_services(self):
         container.stop(Config.SERVICE_NAME)
 
         container.add_layer("mongod", self._mongod_layer, combine=True)
+        if self.is_role(Config.Role.CONFIG_SERVER):
```
Comment (Gu1nness) [inline src/charm.py:1209]: Restart `mongos` when we restart services if we are config server

File: src/charm.py [hunk 8]
```diff
@@ -721,8 +727,17 @@ def _on_update_status(self, event: UpdateStatusEvent):
         # Cannot check more advanced MongoDB statuses if mongod hasn't started.
         with MongoDBConnection(self.mongodb_config, "localhost", direct=True) as direct_mongo:
             if not direct_mongo.is_ready:
-                self.status.set_and_share_status(WaitingStatus("Waiting for MongoDB to start"))
-                return
+                # edge case: mongod will fail to run if 1. they are running as shard and 2. they
```
Comment (Gu1nness) [inline src/charm.py:730]: Edge case (copied from mongodb vm)

File: src/charm.py [hunk 9]
```diff
@@ -370,7 +370,7 @@ def db_initialised(self, value):
             self.app_peer_data["db_initialised"] = str(value)
         else:
             raise ValueError(
-                f"'db_initialised' must be a boolean value. Proivded: {value} is of type {type(value)}"
+                f"'db_initialised' must be a boolean value. Provided: {value} is of type {type(value)}"
```
Comment (Gu1nness) [inline src/charm.py:373]: Fix typo

File: src/charm.py [hunk 10]
```diff
@@ -693,9 +693,15 @@ def _relation_changes_handler(self, event) -> None:
                 self.client_relations.update_app_relation_data()
 
             except NotReadyError:
+                self.status.set_and_share_status(
```
Comment (Gu1nness) [inline src/charm.py:696]: Missing status (compared to MongoDB  VM)

File: tests/integration/sharding_tests/test_sharding_tls.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
+#!/usr/bin/env python3
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/test_sharding_tls.py]: nit:

I know in the VM charm this test is in the directory:
`tests/integration/sharding_tests`

but this scheme is not ideal, instead can you move this test to:
`tests/integration/tls_tests/test_sharding_tls.py`

File: tests/integration/sharding_tests/test_sharding_tls.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+TIMEOUT = 15 * 60
+
+
+@pytest.mark.group(1)
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest) -> None:
+    """Build and deploy a sharded cluster."""
+    await deploy_cluster_components(ops_test)
+
+    # deploy the s3 integrator charm
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/test_sharding_tls.py]:
```suggestion
    # deploy the self-signed-certificates charm
```

File: tests/integration/sharding_tests/test_sharding_tls.py (hunk 11 above)
Comment (Gu1nness) [reply tests/integration/sharding_tests/test_sharding_tls.py]: Sure

File: tests/integration/sharding_tests/test_sharding_tls.py (hunk 12 above)
Comment (Gu1nness) [reply tests/integration/sharding_tests/test_sharding_tls.py]: Done

**Review by Mehdi-Bendriss — APPROVED** [review]
Thank you Neha!

**Review by MiaAltieri — APPROVED** [review]

**Pull request merged**